source/modulo_components/include/modulo_components/ComponentInterface.hpp

```diff
@@ -784,17 +808,48 @@ inline void ComponentInterface<NodeT>::set_predicate(
 }
 
 template<class NodeT>
-inline std::string ComponentInterface<NodeT>::validate_input_signal_name(const std::string& signal_name) {
+inline void ComponentInterface<NodeT>::declare_signal(
+    const std::string& signal_name, const std::string& type, const std::string& default_topic, bool fixed_topic
+) {
   std::string parsed_signal_name = utilities::parse_topic_name(signal_name);
   if (parsed_signal_name.empty()) {
     throw exceptions::AddSignalException(
-        "The parsed signal name for input '" + signal_name
+        "The parsed signal name for " + type + " '" + signal_name
             + "' is empty. Provide a string with valid characters for the signal name ([a-zA-Z0-9_]).");
   }
   if (this->inputs_.find(parsed_signal_name) != this->inputs_.cend()) {
-    throw exceptions::AddSignalException("Input with name '" + parsed_signal_name + "' already exists.");
+    throw exceptions::AddSignalException("Signal with name '" + parsed_signal_name + "' already exists as input.");
+  }
+  if (this->outputs_.find(parsed_signal_name) != this->outputs_.cend()) {
+    throw exceptions::AddSignalException("Signal with name '" + parsed_signal_name + "' already exists as output.");
+  }
+  std::string topic_name = default_topic.empty() ? "~/" + parsed_signal_name : default_topic;
+  auto parameter_name = parsed_signal_name + "_topic";
+  // TODO is the is_empty condition still valid?
```

Comment:
Hmm. `declare_signal` should be done before `add_signal` in the lifecycle case, and manually adding a signal topic parameter should be avoided. Therefore it's unlikely that the parameter would already exist. If it does already exist and has some meaningful value, it definitely shouldn't be set to default (otherwise the sequence of "declare_signal on construction", "user sets topic parameter", "add_signal on configure" would overwrite the user parameter back to default). However, if the parameter exists but is empty, that is a different case.

If someone _really_ wants to manually add the signal topic parameter _and_ leave it empty, then this code would set the signal topic to be the default when the signal is added. I think that's the safest behaviour. The alternative is to not set the parameter value if the parameter already exists, and then in add_input / create_output you'd have to check that the parameter is not empty and the topic name is not empty. It's more work and doesn't bring much apart from the niche feature of "forbidding signals to be created if the parameter hasn't been manually set to be non-empty"

Long story short, let's leave this check in here.

Reply:
What about the scenario where someone declares a signal
```
delcare_signal("test"); // parameter test_topic will have value ~/test
```
and then add the signal with a default topic name later on
```
add_signal("test", "this_topic", ...); // parameter test_topic is not empty and will keep value ~/test
```
Do you think that is an unrealistic scenario or just an incorrect usage?

Reply:
I see that as incorrect usage. There's not a good way of distinguishing if an existing topic parameter value was set by declare_parameter or by a user setting the parameter. Therefore calling add_signal with some different default topic should not apply the new topic. It's a default topic, a.k.a. fallback topic, for the case where the parameter does not exist (or, does not have a value)